pallets/pips/src/benchmarking.rs

```diff
@@ -0,0 +1,377 @@
+        let call = Call::<T>::prune_proposal(0);
+    }: {
+        call.dispatch_bypass_filter(origin)?;
+    }
+    verify {
+        assert_eq!(false, Proposals::<T>::contains_key(&0));
+        assert_eq!(false, ProposalMetadata::<T>::contains_key(&0));
+    }
+
+    reschedule_execution {
+        // length of the proposal padding
+        let c in 0 .. 100_000;
+
+        let (proposer_account, proposer_origin, proposer_did) = make_account::<T>("proposer", 0);
+        identity::CurrentDid::put(proposer_did);
+        let (proposal, url, description) = make_proposal::<T>(c as usize);
+        Module::<T>::set_proposal_cool_off_period(RawOrigin::Root.into(), 0.into())?;
+        Module::<T>::propose(
+            proposer_origin.clone().into(),
+            proposal,
+            42.into(),
+            Some(url.clone()),
+            Some(description.clone())
+        )?;
+        T::GovernanceCommittee::bench_set_release_coordinator(proposer_did);
+        Module::<T>::snapshot(proposer_origin.clone().into())?;
+        let enact_origin = T::VotingMajorityOrigin::successful_origin();
+        let enact_call = Call::<T>::enact_snapshot_results(vec![(0, SnapshotResult::Approve)]);
+        enact_call.dispatch_bypass_filter(enact_origin)?;
+        let future_block = frame_system::Module::<T>::block_number() + 100.into();
+    }: _(proposer_origin, 0, Some(future_block))
+    verify {
+        assert_eq!(true, PipToSchedule::<T>::contains_key(&0));
+    }
+
+    clear_snapshot {
+        // length of the proposal padding
+        let c in 0 .. 100_000;
+
+        let (account, origin, did) = make_account::<T>("proposer", 0);
+        identity::CurrentDid::put(did);
+        let (proposal, url, description) = make_proposal::<T>(c as usize);
+        Module::<T>::set_proposal_cool_off_period(RawOrigin::Root.into(), 0.into())?;
+        Module::<T>::propose(
+            origin.clone().into(),
+            proposal,
+            42.into(),
+            Some(url.clone()),
+            Some(description.clone())
+        )?;
+        T::GovernanceCommittee::bench_set_release_coordinator(did);
+        Module::<T>::snapshot(origin.clone().into())?;
+        assert!(SnapshotMeta::<T>::get().is_some());
+    }: _(origin)
+    verify {
+        assert!(SnapshotMeta::<T>::get().is_none());
+    }
+
+    snapshot {
+        // length of the proposal padding
+        let c in 0 .. 100_000;
+
+        let (account, origin, did) = make_account::<T>("proposer", 0);
+        identity::CurrentDid::put(did);
+        let (proposal, url, description) = make_proposal::<T>(c as usize);
+        Module::<T>::set_proposal_cool_off_period(RawOrigin::Root.into(), 0.into())?;
+        Module::<T>::propose(
+            origin.clone().into(),
+            proposal,
+            42.into(),
+            Some(url.clone()),
+            Some(description.clone())
+        )?;
+        T::GovernanceCommittee::bench_set_release_coordinator(did);
+    }: _(origin)
```

Comment:
The complexity of a snapshot depends on the number of proposals and votes. We need to add a benchmark variable around both to benchmark the actual effect and then we need to take a call on how to represent it in the weight.

Reply:
Currently I'm following the approach of exactly one weight function per extrinsic that is intended to capture the worst-case scenario and without weight adjustment in post-dispatch. Since `snapshot` doesn't take arguments apart from the origin, it will only be possible to apply different weights based on the internal function complexity if we allow adjusting the weight in post-dispatch.

Reply:
@maxsam4 I've added extra parameters. Please have a look and tell me what you think.

Reply:
I know it's not possible to do what I'm asking in the current API. We have a problem(unbounded complexity) and we need a solution. 

Taking a step back and thinking about the architecture of pips, I see a few ways out of this problem:
1) Make it expensive to create proposals so that people/attackers can not afford to ddos this function by creating a large number of proposals.
2) Make snapshotting a more continuous process. i.e. Instead of doing all of the calculations inside this monolithic function, we maintain an ongoing sorted list of proposals that we update with every vote. The snapshot function then just copies that list to a different place.

@Centril @adamdossa Thoughts?

Reply:
We had a brief discussion on slack, but we'll pencil in some discussion on the standup. :)

Reply:
@Centril  There may not be enough time during the standup. I'd appreciate a summary of your discussion.

Reply:
I think we can make the time, but as a brief summary, I think me, @adamdossa, and @maxsam4 have a consensus among ourselves to make snapshotting more continuous and worry about other voting approaches & whatnot later. But let's make sure everyone else is happy too.

I'll probably switch back to tweaking PIPs in that case.
In the interest of not blocking your PR for too long, perhaps we can land it against the current logic, and make tweaks later.

Reply:
I do think making the snapshot process more continuous is the problem solver and allow dividing the cost to voters. What do you mean by the other voting approaches @Centril ?

Reply:
Like quadratic voting with per-DID/scope aggregation.

Reply:
I think that will be like an advancement in the existing voting methodology, IMO this will be a separate concern.